server/controllers/userAuth.js

```diff
@@ -126,9 +126,72 @@ const signUp = (req, resp, next) => {
 
 const confirmEmail = (req, resp) => {};
 
+const resetPassword = (req, resp, next) => {
+  const resetToken = crypto.randomBytes(32).toString('hex');
+  const resetTokenExpirationDate = new Date().getTime() + 3600000;
+  const { email } = req.body;
+  const sanitizedEmail = sanitize(email);
+  const { isEmail, isEmpty } = validator;
+
+  if (isEmpty(sanitizedEmail) || !isEmail(sanitizedEmail)) {
+    return resp.sendStatus(406);
+  }
+
+  User.findOne({ email: sanitizedEmail })
+    .exec()
+    .then(user => {
+      if (!user) {
+        throw new Error('authorization.actions.no-user-of-email');
+      }
+
+      const { displayName, isActive, idFromProvider } = user;
+
+      if (idFromProvider) {
+        throw new Error('authorization.actions.google-user-no-reset-password');
+      }
+
+      if (!isActive) {
+        throw new Error(
+          'authorization.actions.reset-password-not-active-account'
+        );
+      }
+
+      return User.findOneAndUpdate(
+        {
+          email
+        },
+        { resetToken, resetTokenExpirationDate }
+      )
+        .exec()
+        .then(() => {
+          // eslint-disable-next-line no-param-reassign
+          resp.locales = {
+            displayName,
+            email,
+            resetToken
+          };
+
+          next();
+        })
+        .catch(() => {
```

Comment:
why do you catch and throw new error? What is wrong with the catched one ;)?

Reply:
This is because when the user somehow avoids email validation on the front end, the validation rules from backend will send error and by catching that error here I can show the "ERROR MESSAGE" under the input once again here. 
Without this catch, the user will not get any response.

Reply:
but validation is made before the query is sent DB and if something is wrong with email status 406 is send and nothing else is executed

Reply:
hmm consider: 
```
const resetPassword = (req, resp, next) => {
  const { email } = req.body;
  const sanitizedEmail = sanitize(email);
  const { isEmail } = validator;

  if (!isEmail(sanitizedEmail)) {
    return resp.sendStatus(406);
  }

  User.findOne({ email: sanitizedEmail })
    .exec()
    .then(user => {
      if (!user) {
        throw new Error();
      }

      const { isActive, idFromProvider } = user;

      if (idFromProvider) {
        throw new Error('authorization.actions.reset-successful');
      }

      if (!isActive) {
        throw new Error(
          'authorization.actions.reset-password-not-active-account'
        );
      }

      const resetToken = crypto.randomBytes(32).toString('hex');
      const resetTokenExpirationDate = new Date().getTime() + 3600000;

      return User.findOneAndUpdate(
        {
          email
        },
        { resetToken, resetTokenExpirationDate },
        { new: true }
      ).exec();
    })
    .then(user => {
      const { displayName, email, resetToken } = user;
      // eslint-disable-next-line no-param-reassign
      resp.locales = {
        displayName,
        email,
        resetToken
      };

      next();
    })
    .catch(err => {
      const { message } = err;

      if (message) {
        return resp.status(400).send({ message });
      }

      resp.sendStatus(400);
    });
};
```

Reply:
> hmm consider:
> 
> ```
> const resetPassword = (req, resp, next) => {
>   const { email } = req.body;
>   const sanitizedEmail = sanitize(email);
>   const { isEmail } = validator;
> 
>   if (!isEmail(sanitizedEmail)) {
>     return resp.sendStatus(406);
>   }
> 
>   User.findOne({ email: sanitizedEmail })
>     .exec()
>     .then(user => {
>       if (!user) {
>         throw new Error();
>       }
> 
>       const { isActive, idFromProvider } = user;
> 
>       if (idFromProvider) {
>         throw new Error('authorization.actions.reset-successful');
>       }
> 
>       if (!isActive) {
>         throw new Error(
>           'authorization.actions.reset-password-not-active-account'
>         );
>       }
> 
>       const resetToken = crypto.randomBytes(32).toString('hex');
>       const resetTokenExpirationDate = new Date().getTime() + 3600000;
> 
>       return User.findOneAndUpdate(
>         {
>           email
>         },
>         { resetToken, resetTokenExpirationDate },
>         { new: true }
>       ).exec();
>     })
>     .then(user => {
>       const { displayName, email, resetToken } = user;
>       // eslint-disable-next-line no-param-reassign
>       resp.locales = {
>         displayName,
>         email,
>         resetToken
>       };
> 
>       next();
>     })
>     .catch(err => {
>       const { message } = err;
> 
>       if (message) {
>         return resp.status(400).send({ message });
>       }
> 
>       resp.sendStatus(400);
>     });
> };
> ```

this is 99% the same implementation that I have.